File: RB101_RB109_small_problems/medium_2/triangle_sides.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: three numbers, representing the length of a triangle's sides
-  Output: symbol representing whether the triangle is equilateral, isosceles, scalene, or invalid

---

**Problem Domain:**

Triangles can be classified based on the length of their sides:

equilateral = all three sides are of equal length
isosceles = two sides are of equal length while the 3rd side is differnt
scalene = all threee sides are of different length

A valid triangle has the following properties:
  - all three sides must have lengths greater than 0
  - sum of the shortest two sides' lengths must be greater than the longest side's length

---

**Explicit Requirements:**
- Must return a symbol
- See Problem Domain for further requirements of each type of triangle
---


**Implicit Requirements:**
- Method must be named #triangle
---

**Clarifying Questions:**

1. None yet.

---

**Mental Model:**

Check if a triangle is valid:
  - Check if all three sides are greater than 0
  - Check if the sum of the shortest two sides' length is greater than the longest side's length

Check how many unique side lengths exist.
  - If only 1 unique side, triangle is equilateral.
  - If two unique sides, triangle is isosceles,
  - If three unique sides, triangle is scalene.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

triangle(3, 3, 3) == :equilateral
triangle(3, 3, 1.5) == :isosceles
triangle(3, 4, 5) == :scalene
triangle(0, 3, 3) == :invalid
triangle(3, 1, 1) == :invalid

---

Data Structure
--------------
Use an array to contain and organize side lengths.
---

Algorithm
---------

- Place input numbers into an sorted array. Assign to 'side_lengths'.
- Determine the number of unique sides in 'side_lengths' array. Assign to variable 'unique_side_count'
- If a side is less than or equal to 0 OR the sum of the shortest two sides' length is less than or equal to  the longest side's length
  - Return :invalid
- Based on the number assigned to 'unique_side_count'
  - If 1, return :equilateral
  - If 2, return :isosceles
  - If 3, return :scalene

Code
----

=end

def triangle(num1, num2, num3)
  side_lengths = [num1.to_f, num2.to_f, num3.to_f].sort
  unique_side_count = side_lengths.uniq.size
  
  return :invalid if invalid?(side_lengths)
  case unique_side_count
  when 1 then :equilateral
  when 2 then :isosceles
  when 3 then :scalene
  end
end

def invalid?(side_lengths_arr)
  side_lengths_arr.any? {|length| length <= 0} || 
  side_lengths_arr[0] + side_lengths_arr[1] <= side_lengths_arr[2]
end

p triangle(3, 3, 3) == :equilateral
p triangle(3, 3, 1.5) == :isosceles
p triangle(3, 4, 5) == :scalene
p triangle(0, 3, 3) == :invalid
p triangle(3, 1, 1) == :invalid

File: JS210_small_problems/medium_2/triangle_sides.js
"use strict";
/*

A triangle is classified as follows:

Equilateral: All three sides are of equal length.
Isosceles: Two sides are of equal length, while the third is different.
Scalene: All three sides are of different lengths.
To be a valid triangle, the sum of the lengths of the two shortest sides must be greater than the length of the longest side, and every side must have a length greater than 0. If either of these conditions is not satisfied, the triangle is invalid.

Write a function that takes the lengths of the three sides of a triangle as arguments and returns one of the following four strings representing the triangle's classification: 'equilateral', 'isosceles', 'scalene', or 'invalid'.

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: three numbers representing the three sides of a triangle
-  Output: return a string representing what kind of triangle is represented by the input integers

---

**Problem Domain:**

Equilateral triangle have three sides of equal length.

Isosceles triangele have two sides of equal length and the third side is different

Scalene has three side of different length.

Valid triangle must have sides with lengths greater than 0 AND the shortest two sides must sum to be greater than the third side.

---

**Explicit Requirements:**
- See Problem Domain
- Return a string
  - "equilateral" for equilateral triangles
  - "isosceles" for isosceles triangles
  - "scalene" for scalene triangeles
  - "invalid" for invalid triangles
---


**Implicit Requirements:**
- The function must be named triangle
---

**Clarifying Questions:**

1. Will three arguments always be provided?
  No.
2. Will the arguments always be integers?
  Yes, assume so.
3. So the inputs will never be BigInt types?
  No.
4. Should we ignore any arguments that are passed in beyond the third argument?
  If too few arguments or too many arguments are passed, return "invalid".

---

**Mental Model:**

- validate the input
  - Return the string "invalid" if the input is invalid
- Place the input integers into an array and sort them based on their numeric value.
- Return the appropriate string based on the classificaition of the triangle

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

triangle(3, 3, 3);        // "equilateral"
triangle(3, 3, 1.5);      // "isosceles"
triangle(3, 4, 5);        // "scalene"
triangle(0, 3, 3);        // "invalid"
triangle(3, 1, 1);        // "invalid"
triangle(3);              // "invalid"
triangle(3, 3, 3, 3);     // "invalid"
triangle(3, 3);           // "invalid"
triangle(3, 3, "3");      // "invalid"

---

Data Structure
--------------

Use an array to contain the numbers representing the sides of the triangle.

[3, 3, 3]

---

Algorithm
---------

#triangle
- Place the input arguments into an array and sort them based on their numeric value.
- validate the input
  - Return the string "invalid" if the input is invalid (use #isInvalidSidesArray)
- Return the appropriate string based on the classificaition of the triangle
  - filter the array of arguments to only the unique numbers in the array
  - return "equilateral" if there is only 1 unique number in the array of arguments
  - return "scalene" if there are 3 unique numbers in the array of arguments
  - otherwise, return "isosceles"

#isInvalidSidesArray(sidesArr)
- return true IF
  - the length of the sidesArr is not strictly equal to 3
  - the elements of the sidesArr are not all numbers (i.e. typeof element !== "number" for all elements)
  - any of the numbers in the sidesArr are less than or equal to 0
  - the first two numbers in sidesArr sum to be less than  or equal to the last number in sidesArr
- else return false

Code
----

*/

function isInvalidSidesArray(sidesArr) {
  return sidesArr.length !== 3 ||
            sidesArr.some(element => typeof(element) !== 'number') ||
            sidesArr.some(num => num <= 0) ||
            (sidesArr[0] + sidesArr[1] <= sidesArr[2]);
}

function triangle(...sidesArr) {
  sidesArr.sort((a, b) => a - b);
  if (isInvalidSidesArray(sidesArr)) return "invalid";

  if (sidesArr[0] === sidesArr[2]) {
    return "equilateral";
  } else if (sidesArr[0] === sidesArr[1] || sidesArr[1] === sidesArr[2]) {
    return "isosceles";
  } else {
    return "scalene";
  }
}

console.log(triangle(3, 3, 3));        // "equilateral"
console.log(triangle(3, 3, 1.5));      // "isosceles"
console.log(triangle(3, 4, 5));        // "scalene"
console.log(triangle(0, 3, 3));        // "invalid"
console.log(triangle(3, 1, 1));        // "invalid"
console.log(triangle(3));              // "invalid"
console.log(triangle(3, 3, 3, 3));     // "invalid"
console.log(triangle(3, 3));           // "invalid"
console.log(triangle(3, 3, "3"));      // "invalid"